libraries/bezier-rs/src/subpath/solvers.rs

```diff
@@ -58,6 +63,32 @@ impl Subpath {
 
 		intersection_t_values
 	}
+
+	/// Returns a list of `t` values that correspond to the self intersection points of the subpath. For each intersection point, the returned `t` value is the smaller of the two that correspond to the point.
+	/// - `error` - For intersections with non-linear beziers, `error` defines the threshold for bounding boxes to be considered an intersection point.
+	/// - `minimum_seperation`: the minimum difference 2 adjacent `t`-values must have, when comparing adjacent t-values, if 2 `t`-values do not satisfy the
+	/// condition, the function takes the larger `t`-value
+	pub fn self_intersections(&self, error: Option<f64>, minimum_seperation: Option<f64>) -> Vec<f64> {
+		let mut intersections_vec = Vec::new();
+		let n = self.len_segments();
+		// TODO: optimization opportunity - this for-loop currently compares all intersections with all curve-segments in the subpath collection
+		for i in 0..n {
+			let other = self.iter().nth(i).unwrap();
```

Comment:
It feels a bit counter-intuitive to how iterators are usually used to be doing a for-loop over `i` and then selecting the `i`th element in the iterator. Could you try using `foreach` instead here, and in the loop below? If you still want the indices, you can use `enumerate` method beforehand as well. The iterator `skip` method would also probably work well for how you only iterate from `i` onwards in the second loop structure (e.g. `self.iter().skip(i).foreach(...)`